pkg/image/trigger/deploymentconfigs/deploymentconfigs.go

```diff
@@ -0,0 +1,253 @@
+		switch {
+		case len(oldTriggers) == 0:
+			change = cache.Added
+		case !reflect.DeepEqual(oldTriggers, triggers):
+			change = cache.Updated
+		}
+	}
+
+	if len(triggers) > 0 {
+		key := i.prefix + dc.Namespace + "/" + dc.Name
+		return key, &trigger.CacheEntry{
+			Key:       key,
+			Namespace: dc.Namespace,
+			Triggers:  triggers,
+		}, change, nil
+	}
+	return "", nil, change, nil
+}
+
+// DeploymentConfigReactor converts image trigger changes into updates on deployments.
+type DeploymentConfigReactor struct {
+	Client client.DeploymentConfigsNamespacer
+}
+
+func UpdateDeploymentConfigImages(dc *deployapi.DeploymentConfig, tagRetriever trigger.TagRetriever) (*deployapi.DeploymentConfig, bool, error) {
+	var updated *deployapi.DeploymentConfig
+
+	// copy the object and reset dc to the copy
+	copyObject := func() {
+		if updated != nil {
+			return
+		}
+		copied, err := kapi.Scheme.Copy(dc)
+		if err != nil {
+			return
+		}
+		dc = copied.(*deployapi.DeploymentConfig)
+		updated = dc
+	}
+
+	for i, t := range dc.Spec.Triggers {
+		p := t.ImageChangeParams
+		if p == nil || p.From.Kind != "ImageStreamTag" || len(p.LastTriggeredImage) > 0 {
+			continue
+		}
+		if !p.Automatic {
+			continue
+		}
+
+		namespace := p.From.Namespace
+		if len(namespace) == 0 {
+			namespace = dc.Namespace
+		}
+
+		ref, _, ok := tagRetriever.ImageStreamTag(namespace, p.From.Name)
+		if !ok && len(p.LastTriggeredImage) == 0 {
+			glog.V(4).Infof("trigger %#v in deployment %s is not resolveable", p, dc.Name)
+			return nil, false, nil
+		}
+		if len(ref) == 0 {
+			ref = p.LastTriggeredImage
+		}
+
+		if p.LastTriggeredImage != ref {
+			copyObject()
+			p = dc.Spec.Triggers[i].ImageChangeParams
+			p.LastTriggeredImage = ref
+		}
+
+		for i, c := range dc.Spec.Template.Spec.InitContainers {
+			if !namesInclude(p.ContainerNames, c.Name) || c.Image == ref {
+				continue
+			}
+			copyObject()
+			container := &dc.Spec.Template.Spec.InitContainers[i]
+			container.Image = ref
+		}
+
+		for i, c := range dc.Spec.Template.Spec.Containers {
+			if !namesInclude(p.ContainerNames, c.Name) || c.Image == ref {
+				continue
+			}
+			copyObject()
+			container := &dc.Spec.Template.Spec.Containers[i]
+			container.Image = ref
+		}
+	}
+	return updated, true, nil
+}
+
+// ImageChanged is passed a deployment config and a set of changes.
+func (r *DeploymentConfigReactor) ImageChanged(obj interface{}, tagRetriever trigger.TagRetriever) error {
+	dc := obj.(*deployapi.DeploymentConfig)
+	copied, err := kapi.Scheme.DeepCopy(dc)
+	if err != nil {
+		return err
+	}
+	newDC := copied.(*deployapi.DeploymentConfig)
+
+	updated, resolvable, err := UpdateDeploymentConfigImages(newDC, tagRetriever)
+	if err != nil {
+		return err
+	}
+	if !resolvable {
+		if glog.V(4) {
+			glog.Infof("Ignoring changes to deployment config %s, has unresolved images: %s", dc.Name, printDeploymentTriggers(newDC.Spec.Triggers))
```

Comment:
It doesn't calculate the string (expensive in allocations) unless the log level is actually 4.  If you do it the other way, printDeploymentTriggers is always calculated.